### PR TITLE
Feat: add Size()

### PR DESCRIPTION
**Pull request by jorgemmsilva**
# Description of change

Adds a `Size` function where it makes sense, to implement `SerializableWithSize`

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

added a Size() check in the main serialization test function, so that every object that implements the new interface gets the Size checked against the real serialization. 

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x ] My code follows the contribution guidelines for this project
- [x ] I have performed a self-review of my own code
- [x ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ x] I have added tests using ginkgo that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by luca-moser — CHANGES_REQUESTED** [review]
Please include the length prefix bytes within `Size()` function for slice types (`FeatureBlocks`/`UnlockConditions` etc.) (`sum` should start at that size) instead of writing them on every parent object's `Size()` func.

Also please use the `serializer.UintX` constants instead of magic numbers.

**Comment by jorgemmsilva** [comment]
Good points, will fix. I opted to define length prefixes in the parent objects because that's where they are specified (on de/serialization)

File: feat_block_metadata.go [hunk 1]
```diff
@@ -83,6 +84,11 @@ func (s *MetadataFeatureBlock) Serialize(deSeriMode serializer.DeSerializationMo
 		Serialize()
 }
 
+func (s *MetadataFeatureBlock) Size() int {
+	// data length prefix as uint16 = 2 bytes
+	return util.NumByteLen(byte(FeatureBlockMetadata)) + 2 + len(s.Data)
```
Comment (luca-moser) [inline feat_block_metadata.go:89]:
```suggestion
	return util.NumByteLen(byte(FeatureBlockMetadata)) + serializer.UInt16ByteSize + len(s.Data)
```

File: input.go [hunk 2]
```diff
@@ -54,9 +54,17 @@ func (in *Inputs) FromSerializables(seris serializer.Serializables) {
 	}
 }
 
+func (in Inputs) Size() int {
+	sum := 0
```
Comment (luca-moser) [inline input.go:58]:
```suggestion
	sum := serializer.UInt16ByteSize
```

File: native_token.go [hunk 3]
```diff
@@ -245,6 +253,11 @@ func (n *NativeToken) Serialize(_ serializer.DeSerializationMode, deSeriCtx inte
 		Serialize()
 }
 
+func (n *NativeToken) Size() int {
+	// amount = 32 bytes(uint256)
+	return NativeTokenIDLength + 32
```
Comment (luca-moser) [inline native_token.go:258]:
```suggestion
	return NativeTokenIDLength + serializer.UInt256ByteSize
```

File: output.go [hunk 4]
```diff
@@ -224,6 +224,14 @@ func (o *Outputs) FromSerializables(seris serializer.Serializables) {
 	}
 }
 
+func (o Outputs) Size() int {
+	sum := 0
```
Comment (luca-moser) [inline output.go:228]:
```suggestion
	sum := serializer.UInt16ByteSize
```

File: output_basic.go [hunk 5]
```diff
@@ -185,6 +186,15 @@ func (e *BasicOutput) Serialize(deSeriMode serializer.DeSerializationMode, deSer
 		Serialize()
 }
 
+func (e *BasicOutput) Size() int {
+	// length prefixes are 1 byte
```
Comment (luca-moser) [inline output_basic.go:190]:
```suggestion
```

File: transaction.go [hunk 6]
```diff
@@ -143,6 +144,12 @@ func (t *Transaction) Serialize(deSeriMode serializer.DeSerializationMode, deSer
 		Serialize()
 }
 
+func (t *Transaction) Size() int {
+	return util.NumByteLen(uint32(PayloadTransaction)) +
+		t.Essence.Size() +
+		2 + t.UnlockBlocks.Size() //2 bytes length prefix
```
Comment (luca-moser) [inline transaction.go:150]:
```suggestion
		t.UnlockBlocks.Size()
```

File: transaction_essence.go [hunk 7]
```diff
@@ -261,6 +262,19 @@ func (u *TransactionEssence) Serialize(deSeriMode serializer.DeSerializationMode
...
 }
 
+func (u *TransactionEssence) Size() int {
+	payloadSize := util.NumByteLen(uint32(0))
+	if u.Payload != nil {
+		payloadSize = u.Payload.Size()
+	}
+	return util.NumByteLen(byte(TransactionEssenceNormal)) +
+		util.NumByteLen(u.NetworkID) +
+		2 + u.Inputs.Size() + //2 bytes length prefix
```
Comment (luca-moser) [inline transaction_essence.go:272]:
```suggestion
		u.Inputs.Size() +
```

File: transaction_essence.go [hunk 8]
```diff
@@ -261,6 +262,19 @@ func (u *TransactionEssence) Serialize(deSeriMode serializer.DeSerializationMode
...
+		InputsCommitmentLength +
+		2 + u.Outputs.Size() + //2 bytes length prefix
```
Comment (luca-moser) [inline transaction_essence.go:274]:
```suggestion
		u.Outputs.Size() +
```

File: unlock_block.go [hunk 9]
```diff
@@ -97,6 +97,14 @@ func (o UnlockBlocks) ToUnlockBlocksByType() UnlockBlocksByType {
 	return unlockBlocksByType
 }
 
+func (o UnlockBlocks) Size() int {
+	sum := 0
```
Comment (luca-moser) [inline unlock_block.go:101]:
```suggestion
	sum := serializer.UInt16ByteSize
```

**Review by luca-moser — CHANGES_REQUESTED** [review]

**Review by luca-moser — APPROVED** [review]

**Pull request merged**